README.md
Python client for airt service
================

<!-- WARNING: THIS FILE WAS AUTOGENERATED! DO NOT EDIT! -->

## Docs

For full documentation, Please follow the below link:

- <a href="https://docs.airt.ai" target="_blank">https://docs.airt.ai/</a>

## How to install

If you don’t have the airt library already installed, please install it
using pip.

``` console
pip install airt-client
```

## How to use

To access the airt service, you must first create a developer account.
Please fill out the signup form below to get one:

- <https://bit.ly/3hbXQLY>

After successful verification, you will receive an email with the
username and password for the developer account.

Once you have the credentials, use them to get an access token by
calling
[`Client.get_token`](https://airtai.github.io/airt-client/api_client.html#client.get_token)
method. It is necessary to get an access token; otherwise, you won’t be
able to access all of the airt service’s APIs. You can either pass the
username, password, and server address as parameters to the
[`Client.get_token`](https://airtai.github.io/airt-client/api_client.html#client.get_token)
method or store them in the environment variables
**AIRT_SERVICE_USERNAME**, **AIRT_SERVICE_PASSWORD**, and
**AIRT_SERVER_URL**

In addition to the regular authentication with credentials, you can also
enable multi-factor authentication (MFA) and single sign-on (SSO) for
generating tokens.

To help protect your account, we recommend that you enable multi-factor
authentication (MFA). MFA provides additional security by requiring you
to provide unique verification code (OTP) in addition to your regular
sign-in credentials when performing critical operations.

Your account can be configured for MFA in just two easy steps:

1.  To begin, you need to enable MFA for your account by calling the
    [`User.enable_mfa`](https://airtai.github.io/airt-client/api_user.html#user.enable_mfa)
    method, which will generate a QR code. You can then scan the QR code
    with an authenticator app, such as Google Authenticator and follow
    the on-device instructions to finish the setup in your smartphone.

2.  Finally, activate MFA for your account by calling
    [`User.activate_mfa`](https://airtai.github.io/airt-client/api_user.html#user.activate_mfa)
    and passing the dynamically generated six-digit verification code
    from your smartphone’s authenticator app.

You can also disable MFA for your account at any time by calling the
method
[`User.disable_mfa`](https://airtai.github.io/airt-client/api_user.html#user.disable_mfa)
method.

Single sign-on (SSO) can be enabled for your account in three simple
steps:

1.  Enable the SSO for a provider by calling the
    [`User.enable_sso`](https://airtai.github.io/airt-client/api_user.html#user.enable_sso)
    method with the SSO provider name and an email address. At the
    moment, we only support **“google”** and **“github”** as SSO
    providers. We intend to support additional SSO providers in future
    releases.

2.  Before you can start generating new tokens with SSO, you must first
    authenticate with the SSO provider. Call the
    [`Client.get_token`](https://airtai.github.io/airt-client/api_client.html#client.get_token)
    with the same SSO provider you have enabled in the step above to
    generate an SSO authorization URL. Please copy and paste it into
    your preferred browser and complete the authentication process with
    the SSO provider.

3.  After successfully authenticating with the SSO provider, call the
    [`Client.set_sso_token`](https://airtai.github.io/airt-client/api_client.html#client.set_sso_token)
    method to generate a new token and use it automatically in all
    future interactions with the airt server.

For more information, please check:

- [Tutorial](https://docs.airt.ai/Tutorial/) with more elaborate
  example, and

- [API](https://docs.airt.ai/API/client/Client/) with reference
  documentation.

Here’s a minimal example showing how to use airt services to train a
model and make predictions.

In the below example, the username, password, and server address are
stored in **AIRT_SERVICE_USERNAME**, **AIRT_SERVICE_PASSWORD**, and
**AIRT_SERVER_URL** environment variables.

### 0. Get token

``` python
# Importing necessary libraries
from airt.client import Client, DataSource, DataBlob

# Authenticate
Client.get_token()
```

### 1. Connect data

``` python
# The input data in this case is a CSV file stored in an AWS S3 bucket. Before
# you can use the data to train a model, it must be uploaded to the airt server.
# Run the following command to upload the data to the airt server for further 
# processing.
data_blob = DataBlob.from_s3(
    uri="s3://test-airt-service/ecommerce_behavior_csv"
)

# Display the upload progress
data_blob.progress_bar()

# Once the upload is complete, run the following command to preprocess and 
# prepare the data for training.
data_source = data_blob.to_datasource(
    file_type="csv",
    index_column="user_id",
    sort_by="event_time"
)

# Display the data preprocessing progress
data_source.progress_bar()

# When the preprocessing is finished, you can run the following command to 
# display the head of the data to ensure everything is fine.
print(data_source.head())
```

    100%|██████████| 1/1 [01:00<00:00, 60.62s/it]
    100%|██████████| 1/1 [00:35<00:00, 35.39s/it]

                              event_time event_type  product_id  \
    user_id                                                       
    10300217   2019-11-06 06:51:52+00:00       view    26300219   
    253299396  2019-11-05 21:25:44+00:00       view     2400724   
    253299396  2019-11-05 21:27:43+00:00       view     2400724   
    272811580  2019-11-05 19:38:48+00:00       view     3601406   
    272811580  2019-11-05 19:40:21+00:00       view     3601406   
    288929779  2019-11-06 05:39:21+00:00       view    15200134   
    288929779  2019-11-06 05:39:34+00:00       view    15200134   
    310768124  2019-11-05 20:25:52+00:00       view     1005106   
    315309190  2019-11-05 23:13:43+00:00       view    31501222   
    339186405  2019-11-06 07:00:32+00:00       view     1005115   

                       category_id              category_code  \
    user_id                                                     
    10300217   2053013563424899933                       None   
    253299396  2053013563743667055    appliances.kitchen.hood   
    253299396  2053013563743667055    appliances.kitchen.hood   
    272811580  2053013563810775923  appliances.kitchen.washer   
    272811580  2053013563810775923  appliances.kitchen.washer   
    288929779  2053013553484398879                       None   
    288929779  2053013553484398879                       None   
    310768124  2053013555631882655     electronics.smartphone   
    315309190  2053013558031024687                       None   
    339186405  2053013555631882655     electronics.smartphone   

                                   brand    price  \
    user_id                                         
    10300217                     sokolov    40.54   
    253299396                      bosch   246.85   
    253299396                      bosch   246.85   
    272811580                       beko   195.60   
    272811580                       beko   195.60   
    288929779                      racer    55.86   
    288929779                      racer    55.86   
    310768124                      apple  1422.31   
    315309190  dobrusskijfarforovyjzavod   115.18   
    339186405                      apple   915.69   

                                       user_session  
    user_id                                          
    10300217   d1fdcbf1-bb1f-434b-8f1a-4b77f29a84a0  
    253299396  b097b84d-cfb8-432c-9ab0-a841bb4d727f  
    253299396  b097b84d-cfb8-432c-9ab0-a841bb4d727f  
    272811580  d18427ab-8f2b-44f7-860d-a26b9510a70b  
    272811580  d18427ab-8f2b-44f7-860d-a26b9510a70b  
    288929779  fc582087-72f8-428a-b65a-c2f45d74dc27  
    288929779  fc582087-72f8-428a-b65a-c2f45d74dc27  
    310768124  79d8406f-4aa3-412c-8605-8be1031e63d6  
    315309190  e3d5a1a4-f8fd-4ac3-acb7-af6ccd1e3fa9  
    339186405  15197c7e-aba0-43b4-9f3a-a815e31ade40  

### 2. Train

``` python
# We assume that the input data for training a model includes the client_column
# target_column, and timestamp column, which specify the time of an event.
from datetime import timedelta

model = data_source.train(
    client_column="user_id",
    target_column="event_type",
    target="*purchase",
    predict_after=timedelta(hours=3),
)

# Display model training progress
model.progress_bar()

# Once the model training is complete, call the following method to display
# multiple evaluation metrics to evaluate the model's performance.
print(model.evaluate())
```

    100%|██████████| 5/5 [00:00<00:00, 126.62it/s]

                eval
    accuracy   0.985
    recall     0.962
    precision  0.934

### 3. Predict

``` python
# Finally, you can use the trained model to make predictions by calling the
# method below.
predictions = model.predict()

# Display model prediction progress
predictions.progress_bar()

# If the dataset is small enough, you can use the following method to download 
# the prediction results as a pandas DataFrame.
print(predictions.to_pandas().head())
```

    100%|██████████| 3/3 [00:10<00:00,  3.38s/it]

                  Score
    user_id            
    520088904  0.979853
    530496790  0.979157
    561587266  0.979055
    518085591  0.978915
    558856683  0.977960


CHANGELOG.md
# Release notes

<!-- do not remove -->

## 2023.3.0


### Bugs Squashed

- hide all submodules except for airt.client(rename to start with _) ([#12](https://github.com/airtai/airt-client/issues/12))

- fix tables in the airt client docs (Tutorial) ([#10](https://github.com/airtai/airt-client/issues/10))


## 2023.2.0




## 2023.1.0




## 2022.12.0





airt/_helper.py
# AUTOGENERATED! DO NOT EDIT! File to edit: ../notebooks/API_Helper.ipynb.

# %% auto 0
__all__ = ['ensure_is_instance', 'get_base_url', 'post_data', 'get_data', 'delete_data', 'add_ready_column', 'generate_df',
           'get_values_from_item', 'get_attributes_from_instances', 'dict_to_df', 'standardize_phone_number']

# %% ../notebooks/API_Helper.ipynb 2
from typing import *

# %% ../notebooks/API_Helper.ipynb 3
import os
import textwrap
from types import MethodType

import pandas as pd
import requests

from airt._constant import PROD_URL, SERVER_URL

# %% ../notebooks/API_Helper.ipynb 5
def ensure_is_instance(o: Any, cls: Type):
    """A function to check if the object argument is an instance of the class argument.

    Args:
        o: A python object for which the instance needs to be checked.
        cls: The expected instance of the object argument.

    Raises:
        A TypeError if the object is not an instance of the class type.
    """
    if not isinstance(o, cls):
        raise TypeError(
            f"The parameter must be a {cls} type, but got `{type(o).__name__}`"
        )

# %% ../notebooks/API_Helper.ipynb 7
def _get_json(response: requests.Response) -> Any:
    """A function to validate the status of the response object.

    Args:
        response: The response object that encapsulates the server's response.

    Returns:
        A dictionary of the response body.

    Raises:
        ValueError: If the response code is not in range of 200 - 399.
    """
    if response:
        return response.json()
    else:
        if (
            response.headers.get("content-type") == "application/json"
            and "detail" in response.json()
        ):
            raise ValueError(response.json()["detail"])
        else:
            raise ValueError(response.text)

# %% ../notebooks/API_Helper.ipynb 11
def get_base_url(server: Optional[str]) -> str:
    """Return the base URL for the airt server.

    If the server value is `None`, retrive the value from the environment variable `AIRT_SERVER_URL`.
    If the variable is not set as well, then the default public server will be used.
    """

    return server if server is not None else os.environ.get(SERVER_URL, PROD_URL)

# %% ../notebooks/API_Helper.ipynb 15
def post_data(
    url: str,
    token: Optional[str],
    data: Optional[Dict[str, Any]] = None,
    json: Optional[Dict[str, Any]] = None,
) -> Dict[str, Any]:
    """A function to send a POST request.

    Args:
        url: The URL of the server to which the request needs to be sent.
        data: A Dictionary object to send in the body of the POST request. The data sent in this param will automatically be form-encoded by the request library.
        json: A Dictionary object to send in the body of the POST request. The data sent in this param will automatically be JSON-encoded by the request library.
        token: The unique auth token for the client, obtained via calling the `Client.get_token()` method.
            Set it to `None` in `Client.get_token()` to obtain the token.

    Returns:
        A dictionary that encapsulates the response body.

    Raises:
        ConnectionError: If the server is not reachable.
        ValueError: If the response code is not in range of 200 - 399.
    """
    if token is not None:
        headers = {"Authorization": f"Bearer {token}"}
        response = requests.post(url, json=json, data=data, headers=headers)
    else:
        response = requests.post(url, data=data, json=json)
    return _get_json(response)

# %% ../notebooks/API_Helper.ipynb 18
def get_data(url: str, token: Optional[str]) -> Any:
    """Send a GET request.

    Args:
        url: The URL of the server to which the request needs to be sent.
        token: The unique auth token for the client, obtained via calling the `Client.get_token()` method.

    Returns:
        A dictionary that encapsulates the response body.

    Raises:
        ConnectionError: If the server is not reachable.
        ValueError: If the response code is not in range of 200 - 399.
    """
    headers = {"Authorization": f"Bearer {token}"}
    response = requests.get(url, headers=headers)
    return _get_json(response)

# %% ../notebooks/API_Helper.ipynb 20
def delete_data(url: str, token: Optional[str]) -> Dict[str, Any]:
    """Send a DELETE request.

    Args:
        url: The URL of the server to which the request needs to be sent.
        token: The unique auth token for the client, obtained via calling the `Client.get_token()` method.

    Returns:
        A dictionary that encapsulates the response body.

    Raises:
        ConnectionError: If the server is not reachable.
        ValueError: If the response code is not in range of 200 - 399.
    """
    headers = {"Authorization": f"Bearer {token}"}
    response = requests.delete(url, headers=headers)
    return _get_json(response)

# %% ../notebooks/API_Helper.ipynb 23
def add_ready_column(df: pd.DataFrame) -> pd.DataFrame:
    """Add ready column to the DataFrame

    Args:
        df: A pandas DataFrame with completed_steps and total_steps columns

    Returns:
        A pandas DataFrame with ready column
    """

    df["ready"] = df["completed_steps"] == df["total_steps"]
    return df.drop(columns=["completed_steps", "total_steps"])

# %% ../notebooks/API_Helper.ipynb 25
def generate_df(
    items: Union[Dict[str, Any], List[Dict[str, Any]]], columns: list
) -> pd.DataFrame:
    """Generate a DataFrame based on the items length

    Args:
        items: A list encapsulating the response from an API endpoint that needs to be converted into a DataFrame.
        columns: A list of columns names to be included in the DataFrame.

    Returns:
        A DataFrame with a shape of (items, columns), if the length of the items is > 0, otherwise an empty DataFrame with only columns names.
    """

    if len(items) > 0:
        df = pd.DataFrame(items)[columns]
    else:
        df = pd.DataFrame({c: [] for c in columns})

    return df

# %% ../notebooks/API_Helper.ipynb 27
def get_values_from_item(items: list, value: Optional[str] = None) -> str:
    """Get **values** from items seperated by comma.

    Args:
        items: The item list from the response.
        value: The value to extract from each items.

    Returns:
        The **values** as string seperated by comma. If the tags list is empty, <none> will be returned.
    """

    if len(items) == 0:
        return "<none>"

    if value is None:
        return ", ".join(items)

    return ", ".join([str(i[value]) for i in items])

# %% ../notebooks/API_Helper.ipynb 31
def get_attributes_from_instances(
    ox: List[object], attributes: List[str]
) -> List[Dict[str, Any]]:
    """Extract the **attributes** from the instances.

    Args:
        ox: List of instances.
        attributes: Attributes to extract from the instances as a list

    Returns:
        A list encapsulating the attribute name and value pairs of each instance.
    """

    lists = [{i: getattr(o, i) for i in attributes} for o in ox]
    return lists

# %% ../notebooks/API_Helper.ipynb 33
def dict_to_df(d: Dict[str, Any]) -> pd.DataFrame:
    """Convert the dict into a pandas dataframe

    Args:
        d: Dict containing the data and dtypes

    Returns:
        The pandas dataframe constructed from the dict
    """
    data = d["data"]
    dtypes = d["dtypes"]

    df = pd.DataFrame(
        data=data["data"], index=data["index"], columns=data["columns"]
    ).rename_axis(data["index_names"])

    for k, v in dtypes.items():
        df[k] = df[k].astype(v)

    return df

# %% ../notebooks/API_Helper.ipynb 35
def check_and_append_otp_query_param(relative_url: str, otp: Union[str, None]) -> str:
    """Append the otp query parameter to the relative url if its not None

    Args:
        relative_url: The relative url for the route
        otp: user otp

    Returns:
        The updated relative_url if otp is not None, else returns the original relative_url
    """
    if otp is not None:
        relative_url = (
            (relative_url + f"&otp={otp}")
            if len(relative_url.split("?")) > 1
            else (relative_url + f"?otp={otp}")
        )
    return relative_url

# %% ../notebooks/API_Helper.ipynb 39
def standardize_phone_number(phone_number: str) -> str:
    """Standardize the user's phone number

    This function takes the user's phone number in different formats and converts it into a
    standardized format. For example, the user can enter the phone number as **440123456789, +440123456789,
    00440123456789, +44 0123456789, and (+44) 012 345 6789** to register a UK-based phone number and this function takes the user input and
    converts it into 440123456789 standardized format.

    Args:
        phone_number: The phone number to convert into a standardized format.

    Returns:
        The phone number in a standardized format.
    """
    phone_number = "".join(filter(str.isdigit, phone_number))
    if phone_number.startswith("00"):
        phone_number = phone_number[2:]
    return phone_number

# %% ../notebooks/API_Helper.ipynb 41
def add_example_to_docs(o: Any, example: str):
    """Add the given example to the object

    Args:
        o: an object, typically a function or a class, for which the example needs to be added
        example: The example string to add
    """

    original_doc = o.__doc__

    first_line, everything_else = original_doc.split("\n", 1)
    doc_with_example_added = (
        first_line
        + "\n"
        + textwrap.dedent(everything_else)
        + "\n"
        + textwrap.dedent(example)
    )

    o.__doc__ = doc_with_example_added
